cub/util_type.cuh

```diff
@@ -1186,6 +1202,80 @@ template <> struct NumericTraits<unsigned int> :        BaseTraits<UNSIGNED_INTE
 template <> struct NumericTraits<unsigned long> :       BaseTraits<UNSIGNED_INTEGER, true, false, unsigned long, unsigned long> {};
 template <> struct NumericTraits<unsigned long long> :  BaseTraits<UNSIGNED_INTEGER, true, false, unsigned long long, unsigned long long> {};
 
+
+#if CUB_IS_INT128_ENABLED 
+template <>
+struct NumericTraits<__uint128_t>
+{
+  using T = __uint128_t;
+  using UnsignedBits = __uint128_t;
+
+  static constexpr Category       CATEGORY    = UNSIGNED_INTEGER;
+  static constexpr UnsignedBits   LOWEST_KEY  = UnsignedBits(0);
+  static constexpr UnsignedBits   MAX_KEY     = UnsignedBits(-1);
+
+  static constexpr bool PRIMITIVE = false;
```

Comment:
Is there any difference from `BaseTraits` apart from `PRIMITIVE == false`? What does `PRIMITIVE` mean and what is it used for?

Reply:
`BaseTraits` has many specializations. If the question is related to differences between `BaseTraits<{UN,}SIGNED_INTEGER, true, false, ...>` specializations and `__{u,}int128_t` specializations, there's no difference apart from `PRIMITIVE == false`.

In my understanding, `PRIMITIVE == true` means that the type is fundamental and not part of CUDA extension. For instance, all vector types are marked as not primitive. Since `__{u,}int128_t`is an extension, it falls into the same category. `PRIMITIVE` is also used in the decoupled look-back as a way to distinguish between types that can be combined with a flag to form a single architectural word or not. So, marking `__{u,}int128_t` as primitive would break scan at the moment.